src/connector/src/sink/mongodb.rs

```diff
@@ -0,0 +1,860 @@
+    }
+}
+
+struct UpsertCommandBuilder {
+    coll: String,
+    upserts: Array,
+    deletes: HashMap<Vec<u8>, Document>,
+}
+
+impl UpsertCommandBuilder {
+    fn new(coll: String, capacity: usize) -> Self {
+        Self {
+            coll,
+            upserts: Array::with_capacity(capacity),
+            deletes: HashMap::with_capacity(capacity),
+        }
+    }
+
+    fn add_upsert(&mut self, pk: Document, row: Document) -> Result<()> {
+        let pk_data = mongodb::bson::to_vec(&pk)
+            .map_err(|err| anyhow!(err).context("cannot serialize primary key"))?;
+        // under same pk, if the record currently being upserted was marked for deletion previously, we should
+        // revert the deletion, otherwise, the upserting record may be accidentally deleted.
+        // see https://github.com/risingwavelabs/risingwave/pull/17102#discussion_r1630684160 for more information.
+        self.deletes.remove(&pk_data);
+
+        self.upserts.push(bson!( {
+            "q": pk,
+            "u": row,
+            "upsert": true,
+            "multi": false,
+        }));
+
+        Ok(())
+    }
+
+    fn add_delete(&mut self, pk: Document) -> Result<()> {
+        let pk_data = mongodb::bson::to_vec(&pk)
+            .map_err(|err| anyhow!(err).context("cannot serialize primary key"))?;
+        self.deletes.insert(pk_data, pk);
+        Ok(())
+    }
+
+    fn build(self) -> (Option<Document>, Option<Document>) {
+        let (mut upsert_document, mut delete_document) = (None, None);
+        if !self.upserts.is_empty() {
+            upsert_document = Some(doc! {
+                "update": self.coll.clone(),
+                "ordered": true,
+                "updates": self.upserts,
+            });
+        }
+        if !self.deletes.is_empty() {
+            let deletes = self
+                .deletes
+                .into_values()
+                .map(|pk| {
+                    bson!({
+                        "q": pk,
+                        "limit": 1,
+                    })
+                })
+                .collect::<Array>();
+
+            delete_document = Some(doc! {
+                "delete": self.coll,
+                "ordered": true,
+                "deletes": deletes,
+            });
+        }
+        (upsert_document, delete_document)
+    }
+}
+
+type MongodbNamespace = (String, String);
+
+// In the future, we may build the payload into RawBSON to gain a better performance.
+// The current API (mongodb-2.8.2) lacks the support of writing RawBSON.
+struct MongodbPayloadWriter {
+    schema: Schema,
+    pk_indices: Vec<usize>,
+    default_namespace: Namespace,
+    coll_name_field_index: Option<usize>,
+    ignore_coll_name_field: bool,
+    is_append_only: bool,
+    client: Client,
+    buffered_entries: usize,
+    max_entries: usize,
+    // TODO switching to bulk write API when mongodb driver supports it
+    insert_builder: Option<HashMap<MongodbNamespace, InsertCommandBuilder>>,
```

Comment:
Better to introduce an enum 
```
enum CommandBuilder {
    AppendOnly(HashMap<MongodbNamespace, InsertCommandBuilder>),
    Upsert(HashMap<MongodbNamespace, UpsertCommandBuilder>)
}
```
, and then we don't have to store the `is_append_only` flag and redirect the builder by the flag.